security entitlement support

kunalkushwaha opened the pull request:
This is just a rough draft for review of approach.
[edit] rebased with #560 

What this PR includes:
- entitlement package based on libentitlement project.
- usage of entitlement package for setting security settings.
- rest of flow of passing information of `entitlements` based on PR #560

kunalkushwaha commented:
//cc @AkihiroSuda @tonistiigi

tonistiigi commented on client/llb/state.go:
```diff
@@ -257,6 +258,13 @@ func (s State) Network(n pb.NetMode) State {
 func (s State) GetNetwork() pb.NetMode {
 	return getNetwork(s)
 }
+func (s State) Security(n pb.SecMode) State {
```
Add `RunOption` version as well.

tonistiigi commented on frontend/dockerfile/builder/build.go:
```diff
@@ -40,6 +40,7 @@ const (
 	keyImageResolveMode   = "image-resolve-mode"
 	keyGlobalAddHosts     = "add-hosts"
 	keyForceNetwork       = "force-network-mode"
+	keyForceSecurity      = "force-security-mode"
```
lets not add this yet. Can be a follow-up.

tonistiigi commented on solver/llbsolver/solver.go:
```diff
@@ -269,13 +269,17 @@ func notifyCompleted(ctx context.Context, v *client.Vertex, err error, cached bo
 
-var AllowNetworkHostUnstable = false // TODO: enable in constructor
+var AllowNetworkHostUnstable = false       // TODO: enable in constructor
+var AllowSecurityUnconfinedUnstable = true // TODO: enable in constructor
```
Leave this to false or add the constructor initialization in this PR.

tonistiigi commented on solver/llbsolver/solver.go:
```diff
@@ -269,13 +269,17 @@ func notifyCompleted(ctx context.Context, v *client.Vertex, err error, cached bo
 		out = append(out, entitlements.EntitlementNetworkHost)
 	}
+	if AllowSecurityUnconfinedUnstable {
+		out = append(out, entitlements.EntitlementSecurityUnconfined)
```
Where is this validated? I think it is missing the condition in `ValidateEntitlements()`

tonistiigi commented on executor/oci/spec_unix.go:
```diff
@@ -40,11 +41,17 @@ func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mou
 
-	s.Mounts = GetMounts(ctx,
+	ociProfile := entitlements.NewOCIProfile(s, meta.SecMode.String())
+	ociProfile, err = ociProfile.SecurityUnconfined()
```
This seems to be always called (and why the double call) atm.

kunalkushwaha commented:
other simple alternative could be, build  `unconfined` secomp profile as on lines  `defaultProfile()` defined in containerd and remove whole libentitlement code. 

But at some point if more entitlements are required a sub library to define all entitlements will be easy to manage.

kunalkushwaha commented:
The entitlements package from libentitlement is removed and replaced by a simple function based on `defaultSecompProfile` of containerd.

This function is required to reviewed, if its fit for `security.unconfined` profile.

https://github.com/moby/buildkit/pull/570/commits/fb002f56ade4c729477c99b3d8d48176b3de5dbe#diff-20b4a13522a5dac1bba70cce7b8e4e44R14

tonistiigi commented on executor/oci/spec_unix.go:
```diff
@@ -31,6 +35,15 @@ func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mou
+			opts = append(opts, seccomp.WithDefaultProfile())
+		}
+		if meta.SecMode == pb.SecMode_UNCONFINED {
+			opts = append(opts, entitlements.WithDefaultAdminProfile())
```
This should add all caps, disable seccomp/apparmor completely (no seccompsupported check), remove maskpaths and readonly overrides and change sys mounts to rw.

tonistiigi commented on util/entitlements/security.go:
```diff
@@ -0,0 +1,523 @@
+			s.Process.Capabilities.Inheritable = append(s.Process.Capabilities.Inheritable, cap)
+			s.Process.Capabilities.Permitted = append(s.Process.Capabilities.Permitted, cap)
+		}
+		s.Linux.Seccomp = DefaultAdminProfile(s)
```
For the unconfied, we can just skip it and not set seccomp at all. For `security.admin` (future) we would use this profile.

kunalkushwaha commented on solver/llbsolver/solver.go:
```diff
@@ -269,13 +269,17 @@ func notifyCompleted(ctx context.Context, v *client.Vertex, err error, cached bo
 
-var AllowNetworkHostUnstable = false // TODO: enable in constructor
+var AllowNetworkHostUnstable = false        // TODO: enable in constructor
+var AllowSecurityUnconfinedUnstable = false // TODO: enable in constructor
```
I am working on this, shall update the PR on Monday.

Facing issue while `buildctl build --allow=security.unconfined` is passed and `security.confined` is also set by default.

kunalkushwaha commented:
Have updated the code with test case and `buildkitd` flags to enable unsecured entitlements like `network.host` and `security.unconfined`.

There is missing something for both `network.host` & `security.unconfined`, even from client or from code `llb.Security(llb.Security(llb.SecurityModeUnconfined))` is enabled, the Executor always receive default values i.e. `security.confined` & `network.host`.

Also, why [`dispatchRun()`](https://github.com/moby/buildkit/blob/49c29176fe4653a15db7922ec9af772dc2ea6db8/frontend/dockerfile/dockerfile2llb/convert.go#L577) is never invoked. 
Please help on this.

tonistiigi commented on cmd/buildkitd/main.go:
```diff
@@ -207,6 +212,18 @@ func main() {
+		for _, e := range ents {
+			switch e {
+			case "security.unconfined":
+				llbsolver.AllowSecurityUnconfinedUnstable = true
```
Don't use these variables anymore. Instead, add them to a config array and pass to the solver constructor.

tonistiigi replied:
@kunalkushwaha

tonistiigi commented on cmd/buildkitd/main.go:
```diff
@@ -207,6 +212,18 @@ func main() {
 
 		controller.Register(server)
 
+		ents := c.GlobalStringSlice("allow-insecure-entitlement")
```
Add this to config and override with the flag.

tonistiigi replied:
@kunalkushwaha

tonistiigi commented on executor/executor.go:
```diff
@@ -18,6 +18,7 @@ type Meta struct {
 	ReadonlyRootFS bool
 	ExtraHosts     []HostIP
 	NetMode        pb.NetMode
+	SecMode        pb.SecMode
```
`SecurityMode`

tonistiigi commented on util/entitlements/security.go:
```diff
@@ -0,0 +1,523 @@
+// Note: must follow the setting of process capabilities
+func WithDefaultAdminProfile() oci.SpecOpts {
+	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
+		addCaps := []string{
```
This is still a limited list. For unconfined use all caps(different from security.admin).

tonistiigi replied:
ping @kunalkushwaha 

And no seccomp for unconfined.

tonistiigi commented:
@kunalkushwaha Ping. Did my answer clear things up or did I miss anything?

kunalkushwaha commented:
@tonistiigi Sorry for delay. 
Yes its clear, was working on some other stuff so just got delayed. Will update the PR soon.

_PS: I will be on leave for 2 weeks from tomorrow._

AkihiroSuda commented:
@kunalkushwaha What's current status?

kunalkushwaha commented:
Hi,
Sorry for delay in resuming the work.

I need some help in understanding how the entitlements can passed from client (`buildctl build`) workers of `buildkitd`

Current approach in this PR.
1. The entitlement information in passed with `--allow` flag e.g `buildctl build --allow security.unconfined `
2. store entitlements information in [`SolveOpt`](https://github.com/moby/buildkit/blob/master/client/solve.go#L39)
3. This is further passed `ControleClient` [`SoleveRequest`](https://github.com/moby/buildkit/blob/master/api/services/control/control.proto#L65) in function [`client.solve()`](https://github.com/moby/buildkit/blob/master/client/solve.go#L144)
4. similar to step 3, the entitlements is passed to `solver` with [`frontend.SolveRequest`](https://github.com/moby/buildkit/blob/master/control/control.go#L230)
5.  Problem arises here, when [`llbbridge.Solve()`](https://github.com/moby/buildkit/blob/master/solver/llbsolver/bridge.go#L103) passes only `frontendOpt` of `SolveRequest`
 - When control comes to [`frontend/builder/build.Build()`](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/builder/build.go#L51) via `GatewayForwarder.Solve` the options passed from client can be retrieved through `BuildOpts()` i.e. `Opts ` and `LLBCaps` and rest of information is lost.

so currently [`defaultSecurityMode`](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/builder/build.go#L80) have only default `security.confined` & `network.none` entitlements.

I want to use the [`LLBCaps`](https://github.com/moby/buildkit/blob/master/solver/pb/caps.go#L11) as it will be defined for both network and security entitlements.

The point where I am stuck is , where to set these LLBCaps, i.e. `CapExecMetaSecurity`, So this can be access in `builder/build.Buid()` function?

@tonistiigi @AkihiroSuda

kunalkushwaha commented:
@tonistiigi I was  retrieve the value of `entitlements` in `executor.Exec()` functions. I am not getting what should be ideal way of retrieving. 

If possible, please give me a overview of control flow how it goes once control comes to buildkitd for any build command.

tonistiigi commented:
The value you need in exec is not entitlements that is passed with the build and enabled from daemon. What you need there is only the `SecMode` that is defined in the `Meta` and available to you. The validation that this secmode value is allowed to be used with current entitlement settings has already happened before that.

There are 2 sets of values here. One is entitlements that define the allowed actions a current build can perform. The other is the actual action that will be done for specific vertexes. The entitlements only affects the validation of the actual actions. Eg. if you just pass entitlements but don't change your vertexes, nothing would change in your build.

kunalkushwaha commented:
_Note_ : I have updated the intermediate code to explain the issue, it contains too many debug code, please ignore that.

-  I understand,  [`Meta.SecurityMode` ](https://github.com/moby/buildkit/pull/570/commits/2472d41699e3101d7bec04bd4d60edd5e942d4f9#diff-b6e1be4db0a6b8d1aada0afaaa76287dR23), is required to set which will take care of [actual usage](https://github.com/moby/buildkit/pull/570/commits/2472d41699e3101d7bec04bd4d60edd5e942d4f9#diff-c4cd170d5b8f26578babdc3e9d7e6943R42) of  entitlement while executing the RUN command or every worker thread.

Now question is how this `meta.Security` is set.
- Entitlement set is properly set to Job [as you suggested.](https://github.com/moby/buildkit/pull/570/commits/2472d41699e3101d7bec04bd4d60edd5e942d4f9#diff-d2a8a37b660c457a954dfeae935c476eR105) 
- I also added entitlement info to [vertex opts]https://github.com/moby/buildkit/pull/570/commits/2472d41699e3101d7bec04bd4d60edd5e942d4f9#diff-2d535f653ea6b1e3c7f47503d45a2391R159), So I may be able to retrieve it later, but couldn't find place,  where this info can be retrieved and set to `meta`?

Why I said frontend before?
- As the `Exec()` of executor is invoked from `frontendGateway.Solve) --> llbbridge.Exec()` & `meta` is set using `frontendOpts` only. 

I know I am missing something, which I want to know, how this `meta` can be set for security & network.

kunalkushwaha commented:
ping @tonistiigi PTAL

kunalkushwaha commented:
> Don't understand this. The first PR shouldn't need any changes to frontend code. When we add Dockerfile support later it will likely be smth like RUN --security=unconfined cmd and not frontend opts.

Got it. I was trying to test it using dockerfile and thought  `buildctl build --allow security.unconfined .. ` should be able to pass the information. So it wont. It needs separate implementation in front-end as you suggested `RUN --security=unconfined cmd`

kunalkushwaha commented:
@tonistiigi have cleaned up the code. PTAL.

kunalkushwaha commented on executor/oci/spec_unix.go:
```diff
@@ -31,6 +35,11 @@ func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mou
 	}
+	if meta.SecurityMode == pb.SecurityMode_UNCONFINED {
+		opts = append(opts, entitlements.WithDefaultUnconfiedProfile())
+	} else if system.SeccompSupported() && meta.SecurityMode == pb.SecurityMode_UNCONFINED {
```
This was mistake. I thought it was only secomp settings while cleaning code. Now corrected.

kunalkushwaha commented on executor/oci/spec_unix.go:
```diff
@@ -44,8 +53,7 @@ func GenerateSpec(ctx context.Context, meta executor.Meta, mounts []executor.Mou
 	s.Process.Cwd = meta.Cwd
-	s.Process.Rlimits = nil           // reset open files limit
-	s.Process.NoNewPrivileges = false // reset nonewprivileges
+	s.Process.Rlimits = nil // reset open files limit
```
No need to change this. fixed.

kunalkushwaha commented:
Hi @tonistiigi, I have made changes for `config` instead of using variables for enabling `entitlements`, as you suggested. Please check, if changes are okay.

Regarding integration testcase, I have added one [testcase](https://github.com/moby/buildkit/pull/570/files#diff-fce84029d346504e44323c30d21a3808R325) which just test the mode of both `security.confined` & `security.unconfined`. Though currently, `unconfined` test is disabled, as I am not sure how to change the `buildkitd` mode in for particular test. Please help me here.

>  check that processes are running in correct isolation

I had used https://github.com/genuinetools/amicontained initially for testing the `capabilities` of both mode just to verify if `unconfined` mode have all CAPS. by correct isolation you mean same  or something else?

tonistiigi commented:
Yes, amicontained can be a good option for this. Other would be to build detection from source with buildkit itself. I'll take a look at what is the best way for adding the support to integration pkg and get back to you.

kunalkushwaha commented:
> Yes, amicontained can be a good option for this. Other would be to build detection from source with buildkit itself.


I am thinking on test which can be used within build code.
- We can test the CAPS set from Definition e.g
```golang
	st := llb.Image("busybox:latest").
		Run(llb.Shlex(`sh -c 'echo confined'`))

	def, err := st.Marshal()
        def.Metadata[..].Caps[....] // testing this
```
- But this will test the settings, but cannot verify the actual isolation of container that is created.
  - This is not much different then existing test.

- What would be other way to test the container isolation from buildkit code?

justincormack commented on util/entitlements/security.go:
```diff
@@ -0,0 +1,74 @@
+
+// WithUnconfinedSpec sets the default seccomp profile to the spec with Admin capability.
+// Note: must follow the setting of process capabilities
+func WithUnconfinedSpec() oci.SpecOpts {
```
This code should be in a Linux specific file, as all the operations only apply to a Linux spec.

kunalkushwaha replied:
I renamed the file to `secuirty_linux.go`

kunalkushwaha commented:
Hi @tonistiigi , I have updated the testcase for unconfined mode.

the test check for `CapEff` with value being all caps are enabled.
 `cat /proc/self/status | grep CapEff | grep 0000003fffffffff`
```
$ capsh --decode=0000003fffffffff                                                                                   
0x0000003fffffffff=cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read

```
I think this can be merged now if code and design are okay.

tonistiigi commented:
@kunalkushwaha Haven't reviewed yet but the CI seems to be failing on vendor check.

tonistiigi commented on client/client_test.go:
```diff
@@ -322,6 +333,39 @@ func testNetworkMode(t *testing.T, sb integration.Sandbox) {
+	secMode := sb.Value("secmode")
+
+	if secMode != securityUnconfined {
+		t.Skip()
```
We should check that caps are properly restricted in `securityConfined` mode.

kunalkushwaha commented:
Updated the testcase for `SecurityConfined` check and fixed CI issue.

kunalkushwaha commented:
@tonistiigi PTAL. Have fixed the testcase with check for `Confined` & `Unconfined` both. Also CI issue is resolved.

AkihiroSuda commented on util/entitlements/security_linux.go:
```diff
@@ -0,0 +1,74 @@
+		}
+		s.Linux.ReadonlyPaths = []string{}
+		s.Linux.MaskedPaths = []string{}
+		s.Process.ApparmorProfile = ""
```
remove seccomp profile as well?

kunalkushwaha replied:
yes, it was discussed here https://github.com/moby/buildkit/pull/570#discussion_r249166003. 
I am updating function comments too.

AkihiroSuda commented:
sorry needs rebase

kunalkushwaha commented:
not sure why `TestBasicLocalCacheImportExport` failing. This test is not related to this PR


```
    --- FAIL: TestClientIntegration/TestBasicLocalCacheImportExport/worker=containerd-1.0 (11.79s)
        require.go:1159: 
            	Error Trace:	client_test.go:2078
            	            				client_test.go:1458
            	            				client_test.go:1517
            	            				run.go:164
            	Error:      	Should be true
            	Test:       	TestClientIntegration/TestBasicLocalCacheImportExport/worker=containerd-1.0
```

kunalkushwaha commented:
Its rebased and CI error gone.

tiborvass commented on solver/pb/ops.pb.go:
```diff
@@ -54,7 +54,30 @@ func (x NetMode) String() string {
+
+var SecurityMode_name = map[int32]string{
+	0: "CONFINED",
+	1: "UNCONFINED",
```
@justincormack Would you be okay calling `unconfined` `insecure` instead? I know all security profiles call it unconfined, but the user intent feels clearer with `insecure`.

justincormack replied:
Yes I think that makes sense, although confined vs insecure is kind of weird, and I don't think we should call it secure vs insecure...

tiborvass replied:
@justincormack `sandbox`/`sandboxed` ? I'm not too bothered by having both `confined` and `insecure`, because those who will see `confined` will probably have read about its technical meaning, whereas `insecure` should mean something strong to those who won't read much about it.

tiborvass replied:
@justincormack told me offline he prefers `confined` and `insecure`.

@kunalkushwaha would you mind updating `UNCONFINED` to `INSECURE` everywhere in this PR? We can keep the word `CONFINED`.

tonistiigi replied:
btw, I like sandbox/insecure pairing

justincormack replied:
ok with sandbox/insecure too.

kunalkushwaha replied:
@tiborvass ^^

tiborvass replied:
Perfect let’s roll with sandbox/insecure then. Do you mind updating @kunalkushwaha

kunalkushwaha replied:
I will update with sandbox/insecure.

tonistiigi commented:
Now that I think of it, I think it was a mistake to add "default entitlements" https://github.com/moby/buildkit/blob/2453e6f3885f17f0d97ff7c1cc2a03d260d7e28f/util/entitlements/entitlements.go#L21 . I think you should just remove `EntitlementSecurityConfined` cause defining that doesn't make sense in any case. You still need `Security(SecurityModeConfined)` though, just allowing the entitlement is pointless because it is allowed always anyway. All the functionality remains, just remove the constant. Same goes for network-none but if you want I can clean that up myself after this PR is merged.

kunalkushwaha commented:
Update the PR with  changes including.
- `Confined/Unconfined ` changed to `Sandbox/Insecure` https://github.com/moby/buildkit/pull/570#discussion_r269375232
- removed constants for `SecurityConfined & NetworkNone` and there is no default entitlement now. https://github.com/moby/buildkit/pull/570#issuecomment-476420140
- testcase added for error checks for entitlements as suggested in https://github.com/moby/buildkit/pull/570#discussion_r268778393
- moved logic for `sysfs` mount in `r/w mode` in `GenerateSpec()` https://github.com/moby/buildkit/pull/570#discussion_r268848349
